rpc/lib/server/handlers.go

```diff
@@ -113,49 +113,60 @@ func makeJSONRPCHandler(funcMap map[string]*RPCFunc, cdc *amino.Codec, logger lo
 			return
 		}
 
-		var request types.RPCRequest
-		err = json.Unmarshal(b, &request)
+		// first try to unmarshal the incoming request as an array of RPC requests
+		var requests []types.RPCRequest
+		var responses []types.RPCResponse
+		err = json.Unmarshal(b, &requests)
 		if err != nil {
-			WriteRPCResponseHTTP(w, types.RPCParseError(types.JSONRPCStringID(""), errors.Wrap(err, "Error unmarshalling request")))
-			return
-		}
-		// A Notification is a Request object without an "id" member.
-		// The Server MUST NOT reply to a Notification, including those that are within a batch request.
-		if request.ID == types.JSONRPCStringID("") {
-			logger.Debug("HTTPJSONRPC received a notification, skipping... (please send a non-empty ID if you want to call a method)")
-			return
-		}
-		if len(r.URL.Path) > 1 {
-			WriteRPCResponseHTTP(w, types.RPCInvalidRequestError(request.ID, errors.Errorf("Path %s is invalid", r.URL.Path)))
-			return
-		}
-
-		rpcFunc := funcMap[request.Method]
-		if rpcFunc == nil || rpcFunc.ws {
-			WriteRPCResponseHTTP(w, types.RPCMethodNotFoundError(request.ID))
-			return
-		}
-
-		ctx := &types.Context{JSONReq: &request, HTTPReq: r}
-		args := []reflect.Value{reflect.ValueOf(ctx)}
-		if len(request.Params) > 0 {
-			fnArgs, err := jsonParamsToArgs(rpcFunc, cdc, request.Params)
+			// next, try to unmarshal as a single request
+			var request types.RPCRequest
+			err = json.Unmarshal(b, &request)
 			if err != nil {
-				WriteRPCResponseHTTP(w, types.RPCInvalidParamsError(request.ID, errors.Wrap(err, "Error converting json params to arguments")))
+				WriteRPCResponseHTTP(w, types.RPCParseError(types.JSONRPCStringID(""), errors.Wrap(err, "Error unmarshalling request")))
```

Comment:
should we return request.ID instead of `types.JSONRPCStringID("")` here?

Reply:
Will the ID be in there if parsing of `request` fails? I suppose if it isn't an empty string will be returned anyways, so I'll put `request.ID` in there 👍

Reply:
> I suppose if it isn't an empty string will be returned anyways, so I'll put request.ID in there

that's why I'm asking and not just telling you to do so) wanted you to have a look at the code. thank you

Reply:
This test fails now:

```
--- FAIL: TestRPCParams (0.00s)
        Error Trace:    handlers_test.go:84
        Error:      	Not equal:
                    	expected: rpctypes.JSONRPCStringID("")
                    	actual  : <nil>(<nil>)
        Test:       	TestRPCParams
        Messages:   	#2: expected ID not matched in RPCResponse
```

Seems like it results in a `nil` value instead of an empty string, so it's necessary 😁